File: Python/questions/BestTimeToBuyAndSellStockI/best-time-to-buy-and-sell-stock-i.py
import sys

def maxProfit(prices: list[int]) -> int:
        result = 0
        minPrice = sys.maxsize
        
        for price in prices:
            if (price < minPrice):
                minPrice = price
            elif (price > minPrice):
                result = max(result, price - minPrice)
        
        return result

if __name__ == "__main__":
    print(maxProfit([7,1,5,3,6,4]))
    print(maxProfit([7,6,4,3,1]))

File: Java/src/questions/BestTimeToBuyAndSellStockI/BestTimeToBuyAndSellStockI.java
public class BestTimeToBuyAndSellStockI {
    public static int maxProfit(int[] prices) {
        int result = 0;
        int minPrice = Integer.MAX_VALUE;

        for (int price : prices) {
            if (price < minPrice) {
                minPrice = price;
            } else if (price > minPrice) {
                result = Math.max(result, price - minPrice);
            }
        }

        return result;
    }

    public static void main(String[] args) {
        System.out.println(BestTimeToBuyAndSellStockI.maxProfit(new int[] { 7, 1, 5, 3, 6, 4 }));
        System.out.println(BestTimeToBuyAndSellStockI.maxProfit(new int[] { 7, 6, 4, 3, 1 }));
    }
}
